bootloader/grub.go

```diff
@@ -370,5 +376,97 @@ func (g *grub) ManagedAssets() []string {
 //
 // Implements ManagedAssetsBootloader for the grub bootloader.
 func (g *grub) CommandLine(extra []string) (string, error) {
-	return "", fmt.Errorf("not implemented")
+	// we do not trust the on disk asset, use the built-in one
+	assetName := "grub.cfg"
+	if g.recovery {
+		assetName = "grub-recovery.cfg"
+	}
+	staticCmdline, err := staticCommandLineFromGrubAsset(assetName)
+	if err != nil {
+		return "", fmt.Errorf("cannot extract static command line element: %v", err)
+	}
+	bv, err := g.GetBootVars("snapd_extra_cmdline_args")
+	if err != nil {
+		return "", fmt.Errorf("cannot load command line arguments from boot environment: %v", err)
+	}
+	args, err := kernelCommandLineSplit(staticCmdline + " " + bv["snapd_extra_cmdline_args"])
+	if err != nil {
+		return "", err
+	}
+	args = append(args, extra...)
+	// sort arguments so that they match their positions
+	args = sortSnapdKernelCommandLineArgsForGrub(args)
```

Comment:
If the mode args always come first, couldn't we assemble the command line placing extra before the static parameter so that we wouldn't need to sort?

Reply:
I need to think about it. We have: `<static args> <extra args> <mode args>`. Given that we'd like to plug something from the gadget or the args set via `snap set system ..` at some point, those should be listed after the `<static args>` part so that variables can be overridden. However, the mode could be listed first. We will still control what gets added to the command line, so attempts to block/override `snapd_recovery_{mode,system}` would get blocked early.